### ISSUE TITLE
`.set_beta(0)` assigns `weight=NaN` to parameters with zero weight causing plot errors

### ISSUE THREAD

**Issue by lukashergt** [williamjameshandley/anesthetic#84]
**Describe the bug**
When the initial data contains samples with zero weight, these will turn into NaNs when running `.set_beta(0)`. This will then result in errors when trying to plot the prior.

**MWE**

```python
ns = NestedSamples(root="/path/to/anesthetic/tests/example_data/pc")
for i in range(10):
    ns['logL'][i] = -np.inf
    ns['weight'][i] = 0
prior = ns.set_beta(0)
prior.plot_1d(['x0', 'x1'])
```

**Error output**

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-31aeaee3a714> in <module>
----> 1 prior.plot_1d(['x0', 'x1'])

~/.virtualenvs/py38env/lib/python3.8/site-packages/anesthetic/samples.py in plot_1d(self, axes, *args, **kwargs)
    256 
    257         for x, ax in axes.iteritems():
--> 258             self.plot(ax, x, *args, **kwargs)
    259 
    260         return fig, axes

~/.virtualenvs/py38env/lib/python3.8/site-packages/anesthetic/samples.py in plot(self, ax, paramname_x, paramname_y, *args, **kwargs)
    157                     if ncompress is None:
    158                         ncompress = 1000
--> 159                     return kde_plot_1d(ax, self[paramname_x],
    160                                        weights=self.weight,
    161                                        ncompress=ncompress,

~/.virtualenvs/py38env/lib/python3.8/site-packages/anesthetic/plot.py in kde_plot_1d(ax, data, *args, **kwargs)
    322     weights = kwargs.pop('weights', None)
    323     ncompress = kwargs.pop('ncompress', 1000)
--> 324     x, w = sample_compression_1d(data, weights, ncompress)
    325     kde = gaussian_kde(x, weights=w)
    326     p = kde(x)

~/.virtualenvs/py38env/lib/python3.8/site-packages/anesthetic/utils.py in sample_compression_1d(x, w, n)
    344         i = w.index
    345     else:
--> 346         i = numpy.random.choice(w.index, size=n, replace=False, p=w/w.sum())
    347 
    348     # Define sub-samples

mtrand.pyx in numpy.random.mtrand.RandomState.choice()

ValueError: probabilities contain NaN
```

**Expected behavior**

I'd say that `.set_beta` should carry over the zero weight instead of assigning NaNs. Then the `w/w.sum()` in the sample compression wouldn't produce the plotting error. Thoughts?

**Additional context**

Cobaya allows to specify external priors with which you can exclude parts of a multi-parameter space. This will create entries with `log(0)` likelihood and with `0` weight

**Comment by lukashergt** [comment]
A workaround is to create a mask:
```python
mask = np.isfinite(prior.weight)
prior[mask].plot_1d(['x0', 'x1'])
```
which then works fine. However, it would be nicer if this extra step wouldn't be necessary...

**Comment by appetrosyan** [comment]
@williamjameshandley this explains a lot of the errors I was getting. I think NestCheck still has this issue, so should we carry this workaround to #63 or create a PR for nestcheck?

**Comment by williamjameshandley** [comment]
Hi @lukashergt. I think the issue is not the zero weight, but in fact the -np.inf logLs. 

in the function decorated by `@beta.setter`, you can see that there is a line including `self.beta*self.logL`, and it is this which generates the error. The fix to this would be in the same vein as #71, namely to define  `0 * np.inf == 0` in this case (via a `np.where`), as the infinity is a 'logarithmic infinity', and thus subdominant to beta's 0. 

Feel free to submit a PR with this single line fix (ideally with the initially failing test above which then passes with the fix).

One further point is that your line `ns['weight'][i] = 0` doesn't actually do anything, since ns['weight'] is different from ns.weight, and it is the latter which is used for computing the weighted sums. This possible anti-pattern was introduced in #55 and #56. I'm not sure of the best way to ameliorate this clearly confusing state of affairs. One could:
- drop the weight as a column (but this then makes it less clear that it is a weighted frame)
- make the weight column read-only (don't know if this is easy to do in pandas, and would likely cause no end of bother down the line)
- synchronise the weight column with the weight attribute more effectively. This would require someone with an understanding of how the weighted data frames work in `anesthetic/weighted_pandas.py` which is currently just me at the moment (AFAIK).

Would be keen to hear both @lukashergt and @appetrosyan 's views on this.

**Comment by lukashergt** [comment]
> One further point is that your line `ns['weight'][i] = 0` doesn't actually do anything, since ns['weight'] is different from ns.weight, and it is the latter which is used for computing the weighted sums.

What exactly do you mean by "doesn't do anything"? It does change the weight column as can be seen if you do `print(prior.weight)` at the end of my MWE. What is less desirable is that it also changes `ns.weight` and `ns.logL` (because `prior` is a simple copy not a deep copy of `ns`)... 
But in any case, here I'm only doing this for the MWE to create the sample that can be used to show the error.

**Comment by lukashergt** [comment]
> It does change the weight column as can be seen if you do `print(prior.weight)` at the end of my MWE.

Nope, wrong. It actually doesn't change the numbers, but it does change the way it displays the numbers making it look like the weights changed to zero when in fact they are the same numbers as before which were tiny to begin with...

No clue, why the display would change...

`logL` seems to be changed to `-inf`, though, which indeed must be the thing that drives the error.

**Comment by williamjameshandley** [comment]
So the issue is that there is a hidden attribute in the `WeightedSeries` and `WeightedDataFrame` called `_weight`. 

The hidden attribute is necessary for `WeightedSeries`, as if you want the pandas equivalent of a `Series` you can only have one column, so the weights have to be an attribute. This is forwarded as a readable attribute via a `@property`, and the correct behaviour is also enforced for slicing.

The same was originally true for `WeightedDataFrame`, but this was slightly counterintuitive as you would not see the weights in the same way that you see `nlive` or other derived `NestedSamples` columns. The contents of `_weight` is therefore copied into a new column whenever you change `_weight` using the anesthetic functions, but if you go and change this new column manually it has no effect on the weighted calculations (like `mean`, `std`, `median` etc)

**Comment by lukashergt** [comment]
> drop the weight as a column (but this then makes it less clear that it is a weighted frame)

I don't like this solution. It is useful to see the weight column when looking at the full table.

> make the weight column read-only (don't know if this is easy to do in pandas, and would likely cause no end of bother down the line)

Well, it seems like this is already the case, right? Trying to overwrite the weight column actually doesn't work...

**Comment by lukashergt** [comment]
The original issue is closed with #85 , however, I'm not sure whether the discussion about the weight column should be kept open...?

**Comment by williamjameshandley** [comment]
Raised as a new issue #86 for further discussion there
